gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GcsJavaClientWriteChannel.java

```diff
@@ -0,0 +1,179 @@
+      // Try-with-resource will close this end of the pipe so that
+      // the writer at the other end will not hang indefinitely.
+      try (pipeSource) {
+        boolean lastChunk = false;
+        ByteBuffer byteBuffer = ByteBuffer.allocate(MAX_BYTES_PER_MESSAGE);
+        while (!lastChunk) {
+          int remainingCapacity = byteBuffer.remaining();
+          ByteString data =
+              ByteString.readFrom(
+                  ByteStreams.limit(pipeSource, remainingCapacity), remainingCapacity);
+          if (data.size() < remainingCapacity) {
+            lastChunk = true;
+          }
+          byteBuffer.put(data.toByteArray());
+          // switch to read mode
+          byteBuffer.flip();
+          // this could result into partial write
+          writeToGcs(byteBuffer);
+          if (!lastChunk) {
+            // compact buffer for further writing
+            byteBuffer.compact();
+          }
+        }
+        // last chunk could be partially written
+        // uploading all bytes of last chunk
+        if (lastChunk && byteBuffer.hasRemaining()) {
+          while (byteBuffer.hasRemaining()) {
+            writeToGcs(byteBuffer);
+          }
+        }
+        return true;
+      } catch (Exception e) {
+        logger.atSevere().withCause(e).log("Exception while writing to channel");
+        throw new IOException(
+            String.format("Error occurred while uploading resource %s", resourceId.toString()), e);
+      }
+    }
+  }
+
+  private BlobWriteOption[] generateWriteOptions() {
+    List<BlobWriteOption> writeOptions = new ArrayList<>();
+
+    writeOptions.add(BlobWriteOption.disableGzipContent());
+    writeOptions.add(BlobWriteOption.generationMatch());
+    if (createOptions.getKmsKeyName() != null) {
+      writeOptions.add(BlobWriteOption.kmsKeyName(createOptions.getKmsKeyName()));
+    }
+    if (channelOptions.isGrpcChecksumsEnabled()) {
+      writeOptions.add(BlobWriteOption.crc32cMatch());
+    }
+    return writeOptions.toArray(new BlobWriteOption[writeOptions.size()]);
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      super.close();
+      writeChannel.close();
+    } catch (Exception e) {
+      logger.atSevere().withCause(e).log(
+          "Error occurred while closing write channel for resource %s", resourceId.toString());
+      throw new IOException(String.format("Upload failed for '%s'", getResourceString()), e);
+    }
+  }
+
+  @Override
+  public void handleResponse(Boolean response) {
+    this.uploadSuccess = response;
+  }
+
+  @Override
+  protected String getResourceString() {
+    return resourceId.toString();
+  }
+
+  public boolean isUploadSuccessful() {
+    return uploadSuccess;
+  }
+
+  private int writeToGcs(ByteBuffer byteBuffer) throws IOException {
+    int bytesWritten = writeChannel.write(byteBuffer);
+    logger.atFinest().log("Bytes written %d for resource %s", bytesWritten, resourceId.toString());
```

Comment:
nit: can remove .toString(). Will be a bit more efficient, I guess

Reply:
Sure.